### PR TITLE
Add the ability to emit platform wide notifications

### PR DESCRIPTION
**Pull request by cstns**
## Description

Added an admin API that can send platform wide notifications to users that fall into the provided roles.
Added a User model query that retrieves users based on provided roles.

## Related Issue(s)

closes https://github.com/FlowFuse/flowfuse/issues/4554

## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: frontend/src/pages/admin/NotificationsHub.vue
Comment (knolleary) [inline frontend/src/pages/admin/NotificationsHub.vue:23]: Not sure how this is meant to be used for non-external links. It requires quite a deep knowledge of the UI - something self-hosted users won't have. Is there a specific use-case that requires this capability?

**Review by knolleary — COMMENTED** [review]
Overall, looks good.

One pending question about the 'external url' option - more the ability to provide non-external urls. Not sure how self-hosted admins would know what to do with that.

If there isn't a specific need or example where it will be needed, can I suggest we hide for now to avoid confusion?

File: frontend/src/pages/admin/NotificationsHub.vue
Comment (cstns) [reply frontend/src/pages/admin/NotificationsHub.vue:23]: It was mainly for us to know how to properly format the internal url's. I agree there's no use on having it for self-hosted customers. Would leaving the external url as default and hiding the placeholder for the internal routes suffice?

Comment (knolleary) [reply frontend/src/pages/admin/NotificationsHub.vue:23]: The default should be external URL, and I don't think we should have the option for it to be anything else. If we wanted to add support for routing internally, then we can sniff the value to decide if its a url or an object - no need to store that otherwise.

Comment (cstns) [reply frontend/src/pages/admin/NotificationsHub.vue]: I removed the external url selector, forcing all urls to be external

**Comment by cstns** [comment]
The only task left is inserting data into the notifications table. I’ve added a basic implementation that loops through and inserts notifications into the database based on the roles query.

However, this approach could become problematic if there are many users or results, as it might cause a table lock on the notifications table and lead to other issues.

Should we stick with this for now?

**Comment by knolleary** [comment]
> However, this approach could become problematic if there are many users or results, as it might cause a table lock on the notifications table and lead to other issues.

> Should we stick with this for now?

I had a quick look at this part this morning. Had a couple thoughts, but need to poke at it a bit more once other unrelated tasks are complete. Will review tomorrow.

**Review by knolleary — APPROVED** [review]
Going to ship as-is. There are improvements to be made on the scaling of adding notifications to large proportions of users in the database; but for the low-scale usage of the feature we can ship now and improve later.

**Comment by knolleary** [comment]
Decided to hold back; want to validate the scaling issue first. Also, want to add the ability to target team tier as well as role.

Will pick up and complete in next few days.

**Comment by knolleary** [comment]
Have pushed a small refactoring:

1. replaces `recipientRoles` on the API with `filter`. This is an object that can contain different filter options. For this iteration the only property is `roles` - the equivalent of `recipientRoles`. This gives us room to move in the future with other filter options to target different selections of users
2. Added a `bulkSend` api for this use case - which uses `bulkCreate` to create notifications in batches of 200. This is all done under a single transaction which helps reduce load on the DB.

As I write this comment, realise I haven't dbl checked the UI tests for the api change I made... will see how that goes before marking this good to merge.

**Comment by cstns** [comment]
Should we get another set of eyes on the back-end changes? I'm not sure I can/should approve the changes

**Comment by knolleary** [comment]
@cstns I'll give it a fresh look over in the morning, but then we can move forward.

**Comment by knolleary** [comment]
Not working on postgres due to implicit casting between boolean/integer types in the query.

Need to fix and add some tests.

**Comment by knolleary** [comment]
Postgres fixed. Also, for the mock call, changed to use `count` rather than `find` as there's no need to return the actual data.

Finally, added some unit tests for the api side.

**Comment by cstns** [comment]
Should I merge the PR after I give it another quick check?

**Review by knolleary — APPROVED** [review]

**Pull request merged**
